LibraBFT/Impl/IO/OBM/Properties/InputOutputHandlers.agda

```diff
@@ -39,56 +42,73 @@ module handleProposalSpec (now : Instant) (pm : ProposalMsg) where
 
   open handleProposal now pm
 
-  record Contract (pre : RoundManager) (_ : Unit) (post : RoundManager) (outs : List Output) : Set where
-    constructor mkContract
+  record Requirements (pool : SentMessages) (pre : RoundManager) : Set where
     field
-      -- General properties / invariants
-      rmInv              : Preserves RoundManagerInv pre post
-      noEpochChange      : NoEpochChange pre post
-      -- Voting
-      voteAttemptCorrect : Voting.VoteAttemptCorrectWithEpochReq pre post outs (pm ^∙ pmProposal)
-      -- Signatures sent
-      outQcs∈RM          : QC.OutputQc∈RoundManager outs pre -- could also be post, see which is more convenient
-
-  contract : ∀ pre → LBFT-weakestPre (handleProposal now pm) (Contract pre) pre
-  contract pre =
-    epvvSpec.contract pre
-      (RWST-weakestPre-bindPost unit (λ where (myEpoch , vv) → step₁ myEpoch vv) (Contract pre))
-      contract-step₁
-    where
-    contractBail : ∀ outs → OutputProps.NoVotes outs → Contract pre unit pre outs
-    contractBail outs noVotes =
-      mkContract reflPreservesRoundManagerInv (reflNoEpochChange{pre})
-        (Voting.mkVoteAttemptCorrectWithEpochReq (Voting.voteAttemptBailed outs noVotes) tt)
-        outqcs∈pre
-      where
-      postulate -- TODO-1: Prove this (waiting on: updates to RoundManager contracts)
-        outqcs∈pre : QC.OutputQc∈RoundManager outs pre
-
-    contract-step₁ : _
-    proj₁ (contract-step₁ (myEpoch@._ , vv@._) refl) (inj₁ e)  pp≡Left =
-      contractBail _ refl
-    proj₁ (contract-step₁ (myEpoch@._ , vv@._) refl) (inj₂ i) pp≡Left =
-      contractBail _ refl
-    proj₂ (contract-step₁ (myEpoch@._ , vv@._) refl) unit pp≡Right =
-      processProposalMsgMSpec.contract now pm pre (Contract pre) pf
+      mSndr            : NodeId
+      m∈pool           : (mSndr , P pm) ∈ pool
+      qcs∈RmSigsSentB4 : QC.SigsForVotes∈Qc∈Rm-SentB4 pre pool
```

Comment:
I think this (`qcs∈RmSigsSentB4`) should be moved into `rmInv`, since the handler preserves the property between pre and post states that QC signatures in the round manager belong to the pool.

Reply:
Thanks @cwjnkins.  That was my first instinct too, but `rmInv` does not have a pool in scope: it is just about `RoundManager`s.  We could add a pool, but I think it makes sense to keep separate those properties that are purely about `RoundManager`.